testbook/client.py

```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
         code_cell = new_code_cell(lines)
         self.cells.insert(inject_idx, code_cell)
 
-        return TestbookNode(self.execute_cell(inject_idx)) if run else TestbookNode(code_cell)
+        cell = TestbookNode(self.execute_cell(inject_idx)) if run else TestbookNode(code_cell)
 
-    def value(self, name):
-        """Extract a JSON-able variable value from notebook kernel"""
+        if run and pop:
+            self.cells.pop(inject_idx)
 
-        result = self.inject(name)
+        return cell
+
+    def value(self, code, safe=True):
+        result = self.inject(code, pop=True)
         if not self._execute_result(result):
             raise TestbookError('code provided does not produce execute_result')
 
-        code = """
-        from IPython.display import JSON
-        JSON({"value" : _})
+        inject_code = """
+            from IPython.display import JSON
+            JSON({"value" : _})
         """
-        cell = self.inject(code)
-        return cell.outputs[0].data['application/json']['value']
+
+        try:
+            outputs = self.inject(inject_code, pop=True).outputs
+            return outputs[0].data['application/json']['value']
+
+        except ValueError:
+            if not safe:
```

Comment:
Let's try without `safe` first, and raise an exception saying maybe use `ref` method instead

Reply:
I think we should have the option of returning a reference to the output if it cannot be serialized. 
I've made the default for `safe` as `False`.

The `value` method is more of an implementation/helper function, rather than a user-facing function, and we should encourage users to use the `ref` method to create references, and `inject` to inject executable code.

Reply:
I still think I'd rather make the user have to choose value or ref explicitly because the actions I can take after calling `value` are slightly different if I got a reference vs an actual value. But if you feel strongly to keeping it with the default we can try that out.

Reply:
Sure, one way we could do it is by adding an attribute to the error object containing the varname (for the non-JSON serializeable output), and then catch it elsewhere to create a reference to said varname.

Reply:
You could do that. Though I think in general you'll just change your code to call ref instead in that case.